### PR TITLE
Docker-build fails to detect podman

### LINKED ISSUE #41085: Docker-build fails to detect podman
### Describe the bug

I have an application, which uses `quarkus-container-image-docker` extension and I am trying to build the app on a machine with podman. It works for released version of Quarkus, but fails for SNAPSHOT (starting from the 1st of June 2024)

### Expected behavior

Build succeeds

### Actual behavior

```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:build (default) on project code-with-quarkus: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.container.image.docker.deployment.DockerProcessor#dockerBuildFromJar threw an exception: java.lang.IllegalStateException: No container runtime was found. Make sure you have either Docker or Podman installed in your environment.
[ERROR] 	at io.quarkus.deployment.util.ContainerRuntimeUtil.detectContainerRuntime(ContainerRuntimeUtil.java:68)
[ERROR] 	at io.quarkus.deployment.util.ContainerRuntimeUtil.detectContainerRuntime(ContainerRuntimeUtil.java:50)
[ERROR] 	at io.quarkus.container.image.docker.common.deployment.CommonProcessor.lambda$getExecutableName$9(CommonProcessor.java:262)
[ERROR] 	at java.base/java.util.Optional.orElseGet(Optional.java:364)
[ERROR] 	at io.quarkus.container.image.docker.common.deployment.CommonProcessor.getExecutableName(CommonProcessor.java:262)
[ERROR] 	at io.quarkus.container.image.docker.common.deployment.CommonProcessor.buildFromJar(CommonProcessor.java:87)
[ERROR] 	at io.quarkus.container.image.docker.deployment.DockerProcessor.dockerBuildFromJar(DockerProcessor.java:62)
[ERROR] 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:483)
[ERROR] -> [Help 1]

```

### How to Reproduce?

1. `wget -O app.zip "https://code.quarkus.io/d?e=rest&e=container-image-docker&cn=code.quarkus.io"`
2. `unzip -q app.zip && cd code-with-quarkus/`
3. This succeeds: ` mvn clean verify -Dquarkus.platform.version=3.11.0 -Dquarkus.container-image.build=true`
4.  This fails:  `mvn clean verify -Dquarkus.platform.group-id=io.quarkus -Dquarkus.platform.version=999-SNAPSHOT -Dquarkus.container-image.build=true`

### Output of `uname -a` or `ver`

6.7.3-200.fc39.x86_64

### Output of `java -version`

Java version: 21.0.1, vendor: Eclipse Adoptium

### Quarkus version or git rev

6bc64a00d8748e56db2e4bb7a49f13f1c30d51de  (the earliest affected build)

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)

### Additional information

```
$ docker --version
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
podman version 5.0.3
```

### PR DESCRIPTION
**Pull request by edeandrea**
Although I can't reproduce #41085 I found some things which may cause it. See the conversation history in #41085.

Fixes #41085

### CODE REVIEW

**Comment by edeandrea** [comment]
FYI this is an addition to #40729, so it'll have to land wherever that lands.

**Review by cescoffier — CHANGES_REQUESTED** [review]
Following the reported issue, I would add a warning when the docker extension is used and fallback on podman. Thus we inform the user that using the specific podman extension is better.

Now there are teams where some members use docker and some podman. In this case what happens if we use both extensions?

**Comment by edeandrea** [comment]
> Now there are teams where some members use docker and some podman. In this case what happens if we use both extensions?

If they don't care about doing anything podman-specific (like building multi-arch images) then they can continue doing what they are doing today.

If they want to do things which deviate from common functionality (like building multi-arch images) then they would need both extensions & to configure both extensions, and then use `quarkus.container-image.builder=docker|podman` to set which one wins.

**Comment by edeandrea** [comment]
> Following the reported issue, I would add a warning when the docker extension is used and fallback on podman. Thus we inform the user that using the specific podman extension is better

@cescoffier this is done. I've rebased my commits into a single commit.

**Review by cescoffier — APPROVED** [review]

**Comment by gsmet** [comment]
Not sure what's going on but CI looks stuck on this one. I'll see what I can do.

File: core/deployment/src/main/java/io/quarkus/deployment/util/ContainerRuntimeUtil.java [hunk 1]
```diff
@@ -42,21 +43,20 @@ private ContainerRuntimeUtil() {
      * @return a fully resolved {@link ContainerRuntime} indicating if Docker or Podman is available and in rootless mode or not
      * @throws IllegalStateException if no container runtime was found to build the image
      */
-    public static ContainerRuntime detectContainerRuntime() {
-        return detectContainerRuntime(true);
-    }
-
-    public static ContainerRuntime detectContainerRuntime(List<ContainerRuntime> orderToCheckRuntimes) {
+    public static ContainerRuntime detectContainerRuntime(ContainerRuntime... orderToCheckRuntimes) {
```
Comment (ppalaga) [inline core/deployment/src/main/java/io/quarkus/deployment/util/ContainerRuntimeUtil.java:46]: Due to this change, all callers of `detectContainerRuntime()` will fail with NoSuchMethodException unless they are recompiled. Let me send a PR to get the method back.

**Pull request merged**